Server/src/services/vehicle.service.ts

```diff
@@ -200,153 +192,251 @@ export default class VehicleService {
+                // TODO: this could be optimized, e.g. using non-linear weighting and dynamically compute upper bound from last logs
+                let timeWeight = (120 - timePassedSec) / 120
+                timeWeight = timeWeight > 0 ? timeWeight : 0 // (this should not be needed because of filtering above)
+
+                // compute accuracy factor (cut off at 15 meter distance to track)
+                const lastPosition = GeoJSONUtils.parseGeoJSONFeaturePoint(appPositions[i][1].position)
+                if (lastPosition == null) {
+                    // at this point this should not happen anymore
+                    logger.error(`Position ${appPositions[i][1].position} is not in GeoJSON-format, but should be.`)
+                    return null
+                }
+                const projectedPoint = nearestPointOnLine(lineStringData, lastPosition)
+                if (projectedPoint.properties.dist == null) {
+                    // this really should not happen as this is stuff by turf
+                    logger.error(`Somehow turf did not compute distance to track while computing nearest-point-on-line.`)
+                    return null
+                }
+                let accuracy = (15 - projectedPoint.properties.dist * 1000) / 15
+                accuracy = accuracy > 0 ? accuracy : 0
+                weightedPositions.push([predictedTrackKm, timeWeight * accuracy])
+            }
+        }
+
+        // one last check, but this really should not happen as we already checked this (maybe just for future changes)
+        if (weightedPositions.length == 0) {
+            logger.error(`Could not find any position while trying to compute vehicle's position.`)
+            return null
+        }
+
+        // normalize weights (this could probably be done more elegant)
+        let weightSum = 0
+        for (let i = 0; i < weightedPositions.length; i++) {
+            weightSum += weightedPositions[i][1]
+        }
+
+        let avgTrackKm = 0
+        for (let i = 0; i < weightedPositions.length; i++) {
+            weightedPositions[i][1] /= weightSum
+
+            // also calculating average track kilometer here
+            avgTrackKm += weightedPositions[i][0] * weightedPositions[i][1]
+        }
+
+        // in the end we just need to turn the track kilometer into a position again
+        const avgPosition = along(lineStringData, avgTrackKm)
+        GeoJSONUtils.setTrackKm(avgPosition, avgTrackKm)
+        return avgPosition
+    }
+
+    /**
+     * Just a wrapper for getting track position of a vehicle to get its distance along the track.
+     * @param vehicle `Vehicle` to get the distance for
+     * @returns distance of `vehicle` as kilometers along the track, `null` if not possible
+     */
+    public static async getVehicleTrackDistanceKm(vehicle: Vehicle): Promise<number | null>{
+
+        // get track point of vehicle
+        const vehicleTrackPoint = await this.getVehiclePosition(vehicle)
```

Comment:
And here, once again, `getVehiclePosition()` is called, therefore closing the call cycle...